### PR TITLE
Fix dtypes in `cupy.linalg`

### LINKED ISSUE #4309: `cupy.linalg.inv` returns wrong value on integer input
```
>>> cupy.linalg.inv(cupy.array([[1, 2], [3, 4]]))
array([[-inf,  inf],
       [ inf, -inf]])
>>> numpy.linalg.inv(numpy.array([[1, 2], [3, 4]]))
array([[-2. ,  1. ],
       [ 1.5, -0.5]])
>>> cupy.__version__
'9.0.0a1'
```

### PR DESCRIPTION
**Pull request by toslunar**
- Support integer inputs in `cupy.linalg` (close #4309).
- Add `cupy.linalg._util.common_type`: Get dtypes for computation and for NumPy-compatible output.

### CODE REVIEW

**Comment by toslunar** [comment]
This PR is WIP because only `_solve.py` in `cupy/linalg` has been fixed.

**Comment by toslunar** [comment]
TODO: Use `cupy.common_type` for consistency in CuPy.

- `cupy.common_type` is equivalent with `numpy.common_type`.  Then, by default, integer types should use the double precision.
- It's another topic how to configure the default precision for integers in `cupy.common_type`.

**Comment by toslunar** [comment]
`numpy.common_type` rejects `numpy.bool_`, but `numpy.linalg.linalg._commonType` accepts it.
```
>>> numpy.common_type(numpy.array(True))
TypeError: can't get common type for non-numeric array
```

```
>>> x = ~numpy.identity(3, numpy.bool_); x
array([[False,  True,  True],
       [ True, False,  True],
       [ True,  True, False]])
>>> numpy.linalg.matrix_rank(x)
3
```

**Comment by emcastillo** [comment]
Some quick thoughts

> cupy.common_type is equivalent with numpy.common_type. Then, by default, integer types should use the double precision.

I agree with this, our current default precision is float64 unless it is explicitly set.

> It's another topic how to configure the default precision for integers in cupy.common_type.

I wonder if we can add an env var here? `CUPY_DEFAULT_PRECISION`

File: cupy/linalg/_util.py [hunk 1]
```diff
@@ -1,10 +1,17 @@
...
 from numpy import linalg
 
 import cupy
+import cupy._util
 from cupy import core
 import cupyx
 
 
+_default_precision = os.getenv('CUPY_DEFAULT_PRECISION')
+
```
Comment (emcastillo) [inline cupy/linalg/_util.py:13]: Lets add this to the docs :)

**Review by emcastillo — DISMISSED** [review]

**Comment by emcastillo** [comment]
Jenkins, test this please

**Comment by emcastillo** [comment]
Jenkins, test this please

**Comment by toslunar** [comment]
I'll fix the test failures (`cupy_tests/test_cusolver.py::TestSyevj`).

**Comment by toslunar** [comment]
Fixed.

**Comment by emcastillo** [comment]
Jenkins, test this please

**Review by emcastillo — APPROVED** [review]

**Pull request merged**